pkg/checks/external/main.go

```diff
@@ -1265,6 +1269,28 @@ func (ext *Checker) getPodClient() typedv1.PodInterface {
 	return ext.KubeClient.CoreV1().Pods(ext.Namespace)
 }
 
+// GenOwnerRef return the OwnerReference for objects(pod/cr) create by the khcheck checker
+func (ext *Checker) GenOwnerRef() ([]metav1.OwnerReference, error) {
+	khcheck, err := ext.getCheck()
+	if err != nil {
+		return nil, err
+	}
+
+	controller := true
+	blockOwnerDeletion := true
+
+	return []metav1.OwnerReference{
+		metav1.OwnerReference{
+			APIVersion:         CRDGroup + "/" + CRDVersion,
+			Kind:               "KuberhealthyCheck",
+			Name:               khcheck.Name,
+			UID:                khcheck.UID,
+			Controller:         &controller,
+			BlockOwnerDeletion: &blockOwnerDeletion,
+		},
+	}, nil
+}
```

Comment:
I would prefer to have the owner reference target the master Kuberhealthy instance.  This way, if the master pod crashes or gets removed, the checker pods it was managing get cleaned up.

If the CRD is the owner, then the checker pods will never really get cleaned up until the user deletes the check configuration.  The `khcheck` is like a user configuration and the master pod is like the owner, I think.

That said, I think it would make sense to possibly make the `khstate` objects owned by the `khcheck` object that pair to... If you feel up to splitting that out, I think it would be a nice improvement.  If not, we can make an issue to track the idea and flip this over to the master pod being master.

Reply:
If we point the ownerreference of the check pod to the kuberhealth master pod, then when we delete the kuberhealth master pod, the check pod will indeed be cleaned, but the khchecks CR will not be cleared. In any case, after all, only after the creation of khcheck CR can the check pod be created. If you want to get a complete cleanup, I think the relationship chain should be such a kuberhealth master Pod - > khcheck Cr - > khstate Cr - > check pod. I think we should add a logic to point the ownerreference of khcheck CR to kuberhealth master pod. What do you think? If possible, I will add this logic after the holiday.

Reply:
So this is an issue that all operators have, we want to make sure that everything is as clean as possible no matter what the end user does.

But the standard is the CR (Custom Resource) is the owner of the rest of the things that gets created and the controller, aka the operator is just sitting and listening for stuff to do.

@integrii as you might remember my first stab at ownerRef issue was that I wanted to set the khcheck as the ownerRef even if I did a really bad job at it :). And since i didn't have the technical knowledge I agreed to do it the current way since it was better then nothing from my point of view.

Sure if we remove the operator it self we will still have khchecks and khstats left but those are just entries in etcd and it won't take any resources and all of the check pods will go to complete or error.

But on the other hand if we bind the ownerRef to the kuberhealthy master pod sure all of the pods will disappear but this might happen just because the kuberhealthy master pod gets migrated to another node due to the physical node crashes or something else and that will ofc break all the checks and we will get bad data in our monitoring.
And when I think about it, this actually removes the functionality of the secondary operator, since if the master operator ever dies the secondary will never take over with a current check, instead all of the checks will be restarted. Fine the secondary node will take over faster but it's only with a few seconds.

But i definitely agree with you that the khstate should be owned by khcheck.

So in short let khcheck be the owner of the check pod and khstat

 khstat
      |^
khcheck -> checker pod

Reply:
@integrii  I cannot set the ownerref of khcheck in the code because khcheck is created by the user and not by kuberhealthy. I still think it is more reasonable to point the ownerRef of the check pod to khcheck. In other words, it may not be the final state, but it is more convenient to use than the current way.